### PR TITLE
Armor and damage calculation

### PR DESCRIPTION
**Pull request by MaanooAk**
- Added armor attribute (shared)
  - Changed some other attributes shared property
- Changed attack attribute damage variable type
- Created `void UnitAction::damage_object(Unit &target)` which takes the armor from the target, the damage from the entity (stored inside the action) and calculates the actual damage based on [reverse_engineering/damage.md](https://github.com/SFTtech/openage/blob/master/doc/reverse_engineering/damage.md)

(Kept the old `void UnitAction::damage_object(Unit &target, unsigned dmg)` for testing)

### CODE REVIEW

**Comment by MaanooAk** [comment]
The checks fail from: `ImportError: No module named 'buildsystem.codecompliance.modes'` (bug?)
Everything else is ok.

**Comment by TheJJ** [comment]
damnit, didn't commit that file. if you rebase your branch against the current master it will be fixed.

**Comment by MaanooAk** [comment]
Fixed the files, have no idea how the executable bits changed. (Also fixed "syntax"(?) typos at reverse_engineering/damage.md)

File: libopenage/unit/attribute.h [hunk 1]
```diff
@@ -114,6 +115,7 @@ class AttributeContainer {
 
 using attr_map_t = std::map<attr_type, std::shared_ptr<AttributeContainer>>;
 
+using typeamount_set = std::map<int, unsigned int>;
```
Comment (TheJJ) [inline libopenage/unit/attribute.h]: What is that? Please write a docstring. I guess it is the attack class to damage map.

Use `std::unordered_map` please, the `std::map` doesn't hash. And it is not a `set`, is it?

File: libopenage/unit/attribute.h [hunk 2]
```diff
@@ -167,14 +169,44 @@ template<> class Attribute<attr_type::hitpoints>: public AttributeContainer {
...
+	// TODO remove (keep for testing)
 	Attribute(UnitType *type, coord::phys_t r, coord::phys_t h, unsigned int d, UnitType *reset_type)
 		:
 		AttributeContainer{attr_type::attack},
 		ptype{type},
 		range{r},
 		init_height{h},
+		stance{attack_stance::do_nothing},
+		attack_type{reset_type} {
+		damage[4] = d;
```
Comment (TheJJ) [inline libopenage/unit/attribute.h]: what is index 4? if it's a special value that could be named, please use an `enum class`

File: libopenage/unit/attribute.h [hunk 3]
```diff
@@ -167,14 +169,44 @@ template<> class Attribute<attr_type::hitpoints>: public AttributeContainer {
...
+	}
+
+	Attribute(UnitType *type, coord::phys_t r, coord::phys_t h, typeamount_set d, UnitType *reset_type)
```
Comment (TheJJ) [inline libopenage/unit/attribute.h]: we should use a unified constructor for the attributes: the above (old) attribute constructor should directly call this one and create a `typeamount_set` with the one entry. Saves setting all the attributes separately.

File: libopenage/unit/attribute.h (hunk 2 above)
Comment (MaanooAk) [inline libopenage/unit/attribute.h]: The 4 is the `gamedata::hit_class::UNITS_MELEE` which is not exported at the moment. However this constructor is `TODO remove` so no need to change the the whole gamedata exports.

File: libopenage/unit/attribute.h [hunk 4]
```diff
@@ -115,7 +115,11 @@ class AttributeContainer {
 
 using attr_map_t = std::map<attr_type, std::shared_ptr<AttributeContainer>>;
 
-using typeamount_set = std::map<int, unsigned int>;
+/**
+ * An unordered_map with a int key used as a type id
+ * and a unsigned int value used as the amount
+ */
+using typeamount_map = std::unordered_map<int, unsigned int>;
```
Comment (TheJJ) [inline libopenage/unit/attribute.h]: but for what is this map? is it to store "attack/armor class with their corresponding damage/absorbtion value"?

File: libopenage/unit/attribute.h [hunk 5]
```diff
@@ -198,10 +202,10 @@ template<> class Attribute<attr_type::attack>: public AttributeContainer {
 		init_height{h},
 		stance{attack_stance::do_nothing},
 		attack_type{reset_type} {
-		damage[4] = d;
+		damage[4] = d; // 4 = gamedata::hit_class::UNITS_MELEE (not exported at the moment)
 	}
```
Comment (TheJJ) [inline libopenage/unit/attribute.h]: Call `Attribute{type, r, h, {4, d}, reset_type}` as a shortcut here.

File: libopenage/unit/attribute.h (hunk 4 above)
Comment (MaanooAk) [inline libopenage/unit/attribute.h]: Yes, but i didn't write that on the doc because it can be used for other things too.

Comment (TheJJ) [inline libopenage/unit/attribute.h]: ah i see, didn't know and realize that.

**Comment by MaanooAk** [comment]
I have made all changes (no?). Also excellent grammar phrase: 

> [...] each attacker's attack classes is used only if the target has the matching armor class.

**Comment by TheJJ** [comment]
merged with 1bd5703ca6a9caf298778579904644f79610e684

**Pull request closed without merging**